core/ledger/kvledger/snapshot.go

```diff
@@ -44,14 +44,27 @@ type snapshotAdditionalInfo struct {
 	LastBlockCommitHashInHex string `json:"last_block_commit_hash"`
 }
 
-// generateSnapshot generates a snapshot. This function should be invoked when commit on the kvledger are paused
-// after committing the last block fully and further the commits should not be resumed till this function finishes
+// generateSnapshot generates a snapshot. This function should be invoked in the following cases:
+// - after a block is committed, if the smallest snapshot request height matches the block height;
+// - when a snapshot request is submitted, if the request height matches the block height;
+// - when a ledger is opened, if the smallest snapshot request height matches the block height
+//   and such a snapshot has not been generated.
+// In all cases, further commits should not be resumed till this function finishes, which is synchronized
+// by snapshotGenerationRWLock. The caller should acquire snapshotGenerationRWLock.RLock and this function
+// needs to unlock it before it returns. The corresponding snapshot request will be deleted no matter
+// if the snapshot generation is successful or not.
 func (l *kvLedger) generateSnapshot() error {
+	defer func() {
+		l.deleteSnapshotRequest()
+		l.snapshotGenerationRWLock.RUnlock()
```

Comment:
I feel that it will make it more maintainable if you move this line to the goroutine function that invokes `generateSnapshot`. It's always easy to understand lock acquire/release pair at a single place, Then, you won't need the changes caused in the file snapshot_test.go